### PR TITLE
[SPARK-7857][MLLIB] Prevent IDFModel from returning zeros in SparseVe…

### PR DESCRIPTION
**Pull request by karlhigley**
…ctors

When the IDF model's `minDocFreq` parameter is set to a non-zero threshold, the
IDF for any feature below that threshold is set to zero. When the model is used
to transform a set of SparseVectors containing that feature, the resulting
SparseVectors contain entries whose values are explicitly zero. The zero entries
should be omitted in order to simplify downstream processing.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala [hunk 1]
```diff
@@ -218,7 +218,7 @@ private object IDFModel {
           newValues(k) = values(k) * idf(indices(k))
           k += 1
         }
-        Vectors.sparse(n, indices, newValues)
+        Vectors.sparse(n, indices, newValues).toSparse
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Hm, isn't this an optimization to make straight in the `Vectors.sparse` method? it doesn't make sense that you have to call `toSparse` on its output, that that would have any effect.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: I tend to agree that this way feels strange. This was the suggested solution in the JIRA ticket, but I'd be happy to update `Vectors.sparse` instead. I considered it, but wasn't sure it made sense to strip out explicit zeros unless it was actually needed.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Oh, hm, maybe the idea is that checking this condition and 'fixing it' in the constructor is too costly. @jkbradley do I have the logic right?

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: I could imagine updating [this line](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala#L307) in `Vectors.sparse` to something like:
```scala
val (indices, values) = elements.filter(_._2 != 0).sortBy(_._1).unzip
```

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Hm, that's already a fairly heavy operation, so I tend to think it's OK to filter out explicit zeroes. @jkbradley am I missing anything? seems like a win. I'm trying to think of why any code would legitimately rely on an explicit 0.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Making the change above breaks the `foreachActive` [test](https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/mllib/linalg/VectorsSuite.scala#L281), which counts the explicit zeros as active elements. I can't think of why code would rely on explicit zeros in `foreachActive`, but that's where the potential breakage lies. (The rest of the MLlib tests pass.)

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: @dbtsai what do you think about it as it looks like you made the test?
https://github.com/apache/spark/commit/b5d17ef10e2509d9886c660945449a89750c8116
Seems like it should be valid for sparse vectors to ignore 0 entries.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: When we implemented `foreachActive`, we decided to loop through all the `active` elements in the `Vector`; hence, `0` in `Sparse Vector` considers as active. Users have to filter out `zero` explicitly in their code.  I think `Vectors.sparse` should create a sparse vector as the input indices and values array provided by users without filtering out the zero automatically which will be very expensive. 

For this PR, I would recommend change

line 58 to
```
    new IDFModel(idf.compressed)
``` 

and in the transfer function, we implement the two extra cases for idf: sparse and v: sparse, and idf: sparse and v: dense. Thus, we should have better performance. btw, for idf: sparse and v: sparse, we can use array buffer to keep track of non-zero elements.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Makes some sense but `Vectors.sparse` is already doing some expensive operations like a sort. A filter doesn't seem like much extra. 

I get that someone setting 0 directly into a sparse vector might want it to literally record it, but is that someone's expectation in passing it into a factory method?

Anyway maybe you're on to something here, just sounds funny that we have to separately count non-zero elements

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: I can see an argument for consistent behavior with respect to explicit zeros across the three versions of `Vectors.sparse`. Sorting only happens in one of the three (as a way of detecting duplicate identifiers), and while filtering doesn't seem that expensive in comparison, it would be significant extra work in the other two versions.

The suggestion of optimizing performance with `idf.compressed` and extra cases in `transform` makes sense to me, but seems like more than is required to address this particular issue. The thought about `ArrayBuffer` suggests a reasonably small edit to the existing `SparseVector` case, though. I'll push a new commit with that change.

**Comment by karlhigley** [comment]
Updated to use an `ArrayBuffer` instead of `toSparse` to omit explicit zeros in returned SparseVectors.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala (hunk 1 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Most of mllib codebase is using

```scala
def sparse(size: Int, indices: Array[Int], values: Array[Double]): Vector =
```
which will take the raw arrays and construct the sparse vector. The `Seq` version is rarely used.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: That version of the factory method would also work here; it would just require an `unzip`. Curious: why is one preferred over the other?

File: mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala [hunk 2]
```diff
@@ -211,14 +213,16 @@ private object IDFModel {
...
     v match {
       case SparseVector(size, indices, values) =>
-        val nnz = indices.size
-        val newValues = new Array[Double](nnz)
+        val newElements = new ArrayBuffer[(Int, Double)]
         var k = 0
-        while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+        while (k < indices.size) {
+          val newValue = values(k) * idf(indices(k))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Since `idf` can be sparse, lookup will not be constant time operation. You need to make it 4 cases. Also, `indices.size` is not good since in each tight iteration, the `size` method will be called. The following code should work, but I just wrote it here without testing so may have some bugs. You need to add the tests for four different scenarios. 

```scala
    (idf, v) match {
      case (didf: DenseVector, dv: DenseVector) =>
        val didfValues = didf.values
        val dvValues = dv.values

        val newValues = new Array[Double](n)
        var j = 0
        while (j < n) {
          newValues(j) = dvValues(j) * didfValues(j)
          j += 1
        }
        Vectors.dense(newValues)

      case (didf: DenseVector, sv: SparseVector) =>
        val didfValues = didf.values
        val svIndices = sv.indices
        val svValues = sv.values
        val svNnz = svIndices.length

        val newValues = new Array[Double](svNnz)
        var k = 0
        while (k < svNnz) {
          newValues(k) = svValues(k) * didfValues(svIndices(k))
          k += 1
        }
        Vectors.sparse(n, svIndices, newValues)

      case (sidf: SparseVector, dv: DenseVector) =>
        val dvValues = dv.values
        val sidfIndices = sidf.indices
        val sidfValues = sidf.values
        val sidfNnz = sidfIndices.length

        val newValues = new Array[Double](sidfNnz)
        var k = 0
        while (k < sidfNnz) {
          newValues(k) = sidfValues(k) * dvValues(sidfIndices(k))
          k += 1
        }
        Vectors.sparse(n, sidfIndices, newValues)

      case (sidf: SparseVector, sv: SparseVector) =>
        val (largeIndices, largeValues, largeNnz, smallIndices, smallValues, smallNnz) = {
          val sidfIndices = sidf.indices
          val sidfValues = sidf.values
          val sidfNnz = sidfIndices.length

          val svIndices = sv.indices
          val svValues = sv.values
          val svNnz = svIndices.length
          if (sidfNnz > svNnz) {
            (sidfIndices, sidfValues, sidfNnz, svIndices, svValues, svNnz)
          } else {
            (svIndices, svValues, svNnz, sidfIndices, sidfValues, sidfNnz)
          }
        }

        val newIndices = new ArrayBuffer[Int](smallNnz)
        val newValues = new ArrayBuffer[Double](smallNnz)

        var j = 0
        var k = 0
        while (j < smallNnz && k < largeNnz) {
          val smallIndex = smallIndices(j)
          val largeIndex = largeIndices(k)
          
          if (smallIndex == largeIndex) {
            newIndices.append(smallIndex)
            newValues.append(smallValues(j) * largeValues(k))
            j += 1
            k += 1
          } else if (smallIndex > largeIndex) {
            k += 1
          } else {
            j += 1
          }
        }
        Vectors.sparse(n, newIndices.toArray, newValues.toArray)

      case (idfOther, vOther) =>
        throw new UnsupportedOperationException(
          s"Only sparse and dense vectors are supported but got ${idfOther.getClass} " +
          s"for idf vector, and ${vOther.getClass} for term frequence vector.")
    }
```

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: Good point about `indices.size`, I've added a commit to move that outside the loop.

The purpose of this ticket and PR is to remove explicit zeros from the output of `IDFModel.transform`. There may be further performance optimizations to be done in this section of code (e.g. the four cases as you suggest). It's not clear to me that the code in this PR is significantly less performant than what was already there, or that the suggested optimized code addresses the original issue. Maybe I'm missing something?

File: mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala (hunk 1 above)
Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: If the `Array` version of the method is commonly used because it contains some additional conveniences and safety checks, that would align with @srowen's point that the expectations may be different for a factory method. If that's what's driving the preferred usage of the `Array` version, it would tip me toward filtering out explicit zeros there (prior to the sort).

File: mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala [hunk 3]
```diff
@@ -211,14 +213,17 @@ private object IDFModel {
     val n = v.size
     v match {
       case SparseVector(size, indices, values) =>
+        val newElements = new ArrayBuffer[(Int, Double)]
         val nnz = indices.size
-        val newValues = new Array[Double](nnz)
         var k = 0
         while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+          val newValue = values(k) * idf(indices(k))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: `idf(indices(k))` will do binary search when `idf` is `sparse`, and this will be too expensive. Have them handled using specialized method.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: As the diff shows, the existing code is already calling `idf(indices(k))`. That call may indeed be expensive and represent a potential optimization, but it's distinct from the problem this PR is intended to address. Seems like there might be room for a second JIRA/PR to handle the issue you identified.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: The existing code is fine since the idf vector is always dense which has constant time lookup. In your change, that will strongly impact the performance, so I don't think it should belong to a separate PR.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: This change doesn't affect the `idf` vector -- it's still always dense. The existing `transform` code was producing `SparseVector`s containing explicit zeros. This change removes the explicit zeros, but only when the input to `transform` (not the `idf`) is a `SparseVector`.

Comment (karlhigley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala]: If I were to follow the suggestion of updating line 58 to `new IDFModel(idf.compressed)`, then it would be possible for the `idf` vector to be sparse, but I'm not doing that here. That's part of the performance optimization I'm suggesting could happen in a separate PR.

**Comment by dbtsai** [comment]
With `minDocFreq`, the idf vector will result lots of zeros, which means looping through each of elements in idf vector is not efficient. Also, with lots of zeros in idf vector, even the input vector is dense, the transformed result will be sparse. This PR doesn't address this issue. Let's make thing right in one PR.

**Comment by karlhigley** [comment]
I understand the issue you're pointing out, but it hasn't been a practical problem, even with hundreds of thousands of terms. The inclusion of explicit zeros in the output SparseVectors has been a practical problem, which is what led me to submit this JIRA and PR.

This is my first contribution to Spark, and I'm trying to adhere to the contribution guidelines. The guidelines suggest that "simple, targeted" changes are more likely to be accepted than "big bang" changes. It sounds like you're telling me this PR won't be accepted without making an additional optimization which adds ~100 lines of code, requires additional tests, and fixes an issue that was already present in the code before my changes. Is that what you're saying?

**Comment by dbtsai** [comment]
Your PR will impact the performance when idf vector is dense, but the input vector is sparse which is my concern. Of course, we would like to have patch as simple as possible, but we can not have impact on existing behavior. (also, this PR is not as simple as the original description sounds).  We definitely can help you to walk through the process.
